compiler/rustc_middle/src/ty/sty.rs

```diff
@@ -2430,6 +2430,21 @@ impl<'tcx> Ty<'tcx> {
         }
     }
 
+    /// Like [`Ty::to_opt_closure_kind`], but it caps the "maximum" closure kind
+    /// to `FnMut`. This is because although we have three capability states,
+    /// `AsyncFn`/`AsyncFnMut`/`AsyncFnOnce`, we only need to distinguish two coroutine
+    /// bodies: by-ref and by-value.
```

Comment:
Add a "see also ..." comment linking to somewhere where the reason for why we only need to distinguish by-ref and by-value is documented